ykrt/src/trace/swt/patch.rs

```diff
@@ -0,0 +1,134 @@
+use libc::{mprotect, size_t, sysconf, PROT_EXEC, PROT_READ, PROT_WRITE};
+use std::mem;
+use std::{ffi::c_void, sync::Once};
+
+#[cfg(tracer_swt)]
+use crate::yk_trace_basicblock;
+
+#[cfg(tracer_swt)]
+static ORIGINAL_INSTRUCTIONS_INIT: Once = Once::new();
+#[cfg(tracer_swt)]
+static mut ORIGINAL_INSTRUCTIONS: [u8; 1] = [0; 1];
+#[cfg(tracer_swt)]
+static mut PATCH_INSTRUCTIONS: [u8; 1] = [0xC3];
+
+/// This function is used to save the original instructions of a function to .
+///
+/// # Arguments
+///
+/// * `function_ptr` - A usize representing the memory address of the function.
+/// * `instructions` - A mutable pointer to a u8 where the original instructions will be saved.
+/// * `num_of_instructions` - A usize indicating the number of instructions to save.
+///
+#[cfg(tracer_swt)]
+unsafe fn save_original_instructions(
+    function_ptr: usize,
+    instructions: *mut u8,
+    num_of_instructions: usize,
+) {
+    let func_ptr: *const () = function_ptr as *const ();
+    std::ptr::copy_nonoverlapping(func_ptr as *const u8, instructions, num_of_instructions);
+}
+
+/// This function is used to patch a function instructions at runtime.
+///
+/// # Arguments
+///
+/// * `function_ptr` - A usize representing the memory address of the function to be patched.
+/// * `code` - A constant pointer to a u8 vector where the new instructions are located.
+/// * `size` - A size_t indicating the number of bytes to copy from `code`.
+///
+#[cfg(tracer_swt)]
+unsafe fn patch_function(function_ptr: usize, code: *const u8, size: size_t) {
+    let page_size = sysconf(libc::_SC_PAGESIZE) as usize;
+
+    let func_address = ((function_ptr as usize) & !(page_size - 1)) as *mut c_void;
+    let page_size_aligned = (((function_ptr as usize) + mem::size_of_val(&function_ptr))
+        - (func_address as usize)) as usize;
+
+    // Set function memory region to be writable
+    let result = mprotect(
+        func_address,
+        page_size_aligned,
+        PROT_READ | PROT_WRITE | PROT_EXEC,
+    );
+    if result != 0 {
+        panic!("Failed to change memory protection to be writable");
+    }
+
+    // Set function memory region back to be non-writable
+    std::ptr::copy_nonoverlapping(code, function_ptr as *mut u8, size);
+
+    let result = mprotect(func_address, page_size_aligned, PROT_READ | PROT_EXEC);
+    if result != 0 {
+        panic!("Failed to change memory protection to not writable");
+    }
+}
+
+fn is_x86() -> bool {
+    std::env::consts::ARCH == "x86" || std::env::consts::ARCH == "x86_64"
+}
+
+/// This function is used to patch the `yk_trace_basicblock`
+/// function with a single `ret` (0xC3) instruction.
+#[cfg(tracer_swt)]
+pub(crate) unsafe fn patch_trace_function() {
+    if !is_x86() {
```

Comment:
See `codegen/mod.rs` for how to do this (better) at compile-time: guard `mod` with `#[cfg(target_arch = "x86_64")]`.

Reply:
Updated 
👉 86e404b3b7c0c9f633b9c49f748bc65811d2186e
👉 0e88a29e6ccb330d899d869631035d66300a8bd4

Reply:
I don't think we want/need https://github.com/ykjit/yk/commit/86e404b3b7c0c9f633b9c49f748bc65811d2186e: what we want to do is `cdg` either the `mod` statement or only those functions which are arch specific. So https://github.com/ykjit/yk/commit/0e88a29e6ccb330d899d869631035d66300a8bd4 is probably not a bad start.

Reply:
With 86e404b3b7c0c9f633b9c49f748bc65811d2186e I thought to not limit `swt` to x86 so if we run it on other architectures - it will still work just without the runtime patching.
But I'm probably overthinking it cause we support only x86 anyway :)

Reply:
I think this commit makes it more clear why we need conditional compilation here 👉 b2aee3dd12b8144c7e0f1cf3ab7c8be19029950b

Reply:
> With https://github.com/ykjit/yk/commit/86e404b3b7c0c9f633b9c49f748bc65811d2186e I thought to not limit swt to x86 so if we run it on other architectures - it will still work just without the runtime patching.

Right, but I think the thing to do there would be to guard *inside* those functions, otherwise mt.rs has to know about swt's hardware acceleration (or not). b2aee3dd12b8144c7e0f1cf3ab7c8be19029950b seems reasonable enough to me.